### PR TITLE
Report correct max depth in exception message

### LINKED ISSUE #34406: JsonSerializer reports the wrong max depth in exception message
The default value for `JsonSerializationOptions.MaxDepth` is 0, which is a special value meaning "use the default" (which is 64)

But when a cycle is detected, the error message reports that the max depth is 0, which is quite confusing. Since I didn't think I had a cycle in my object graph, I initially thought I had accidentally set the MaxDepth to 0.

> System.Text.Json.JsonException : A possible object cycle was detected which is not supported. This can either be due to a cycle or if the object depth is larger than the maximum allowed depth of 0.

It should probably report the *effective* max depth.

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #34406

### CODE REVIEW

**Review by danmoseley — APPROVED** [review]

**Review by jozkee — APPROVED** [review]
It would be good to also have tests for both code paths.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 1]
```diff
@@ -207,7 +207,7 @@ internal bool TryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions opt
         {
             if (writer.CurrentDepth >= options.EffectiveMaxDepth)
             {
-                ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.MaxDepth);
+                ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.EffectiveMaxDepth);
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:210]: @thomaslevesque - are up for submitting a PR with tests for this? If not, @Jozkee - can you do so (or file an up-for-grabs issue)?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 2]
```diff
@@ -289,7 +289,7 @@ internal bool TryWriteDataExtensionProperty(Utf8JsonWriter writer, T value, Json
 
             if (writer.CurrentDepth >= options.EffectiveMaxDepth)
             {
-                ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.MaxDepth);
+                ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.EffectiveMaxDepth);
             }
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:293]: It looks to me that this condition with never be true; the `JsonExtensionData` dictionary itself does not increase the `CurrentDepth` of the writer, the elements within does it and that is already covered in the `TryWrite` method.

My point here is that stepping into this branch is impossible and instead we should probably remove it.
Hence, having code coverage for this code path (https://github.com/dotnet/runtime/pull/34410/files#r401931623) does not apply after all.

@steveharter @layomia thoughts?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 1 above)
Comment (thomaslevesque) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:210]: I can try, if I figure out how to run the tests...
So far I've been unable to load the solution in VS. Do I need a preview version?

Comment (ahsonkhan) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:210]: @ViktorHofer can you please help @thomaslevesque here or direct him to the latest docs? Tyvm!

Comment (danmosemsft) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:210]: @thomaslevesque what version do you have? Our projects require VS2019 (16.x) to open. Do you have 16.5.0 (see Help>About)? Although, earlier releases of 16.x may work.

Comment (thomaslevesque) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:210]: @ahsonkhan  @danmosemsft I eventually managed to build and run the tests. I just modified an existing test, I hope that's ok. See #34528 
It only tests one of the paths. As mentioned by @Jozkee, the other path seems unreachable.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:293]: This makes sense @Jozkee. Let's open a PR.

**Pull request merged**